File: WEEK 1/DAY 6 - Stacks Basics/SOLUTIONS/03_Hard_Problem_Solution/Solution.cpp
#include <vector>
#include <stack>

class Solution {
public:
    std::vector<int> dailyTemperatures(std::vector<int>& temperatures) {
        int n = temperatures.size();
        std::vector<int> answer(n, 0); // Initialize answer array with 0s
        std::stack<int> s; // Stack to store indices
        
        for (int i = 0; i < n; ++i) {
            while (!s.empty() && temperatures[i] > temperatures[s.top()]) {
                int idx = s.top();
                s.pop();
                answer[idx] = i - idx; // Calculate the number of days
            }
            s.push(i); // Push current day index to stack
        }
        
        return answer;
    }
};

File: WEEK 1/DAY 6 - Stacks Basics/SOLUTIONS/03_Hard_Problem_Solution/Solution.js
/**
 * @param {number[]} temperatures
 * @return {number[]}
 */
var dailyTemperatures = function(temperatures) {
    const n = temperatures.length;
    const answer = new Array(n).fill(0);
    const stack = [];
    
    for (let i = 0; i < n; i++) {
        while (stack.length && temperatures[i] > temperatures[stack[stack.length - 1]]) {
            const idx = stack.pop();
            answer[idx] = i - idx;
        }
        stack.push(i);
    }
    
    return answer;
};
